src/components/tooltip/tooltip.tsx

```diff
@@ -0,0 +1,185 @@
+import * as debounce from 'debounce';
+import * as React from 'react';
+import {stylable} from 'wix-react-tools';
+
+import {GlobalEvent} from '../global-event';
+import {Portal} from '../portal';
+import styles from './tooltip.st.css';
+
+export type Position = 'top' | 'left' | 'right' | 'bottom';
+
+export interface TooltipProps {
+    children?: React.ReactNode;
+    position?: Position;
+    id: string;
+    open?: boolean;
+    showTrigger?: string;
+    hideTrigger?: string;
+    showDelay?: number;
+    hideDelay?: number;
+}
+
+export interface TooltipState {
+    style?: React.CSSProperties;
+    open: boolean;
+}
+
+@stylable(styles)
+class StyledTooltip extends React.Component<TooltipProps, TooltipState> {
+    public static defaultProps = {
+        open: false,
+        position: 'top',
+        showTrigger: 'mouseenter',
+        hideTrigger: 'mouseleave',
+        showDelay: 0,
+        hideDelay: 0
+    };
+
+    private target: HTMLElement | null = null;
+    private events: string[] = [];
+    private timeout?: number;
+    private onWindowResize = debounce(() => {
+        if (this.state.open) {
+            this.setStyles();
+        }
+    }, 500);
+
+    public constructor(props: TooltipProps) {
+        super();
+        this.state = {
+            style: undefined,
+            open: props.open!
+        };
+    }
+
+    public render() {
+        const {children, position} = this.props;
+        const {style, open} = this.state;
+
+        if (!style) {
+            return null;
+        }
+
+        return (
+            <div
+                data-automation-id="TOOLTIP"
+                className={`root ${position}`}
+                style={style}
+                style-state={{open}}
+            >
+                <GlobalEvent
+                    resize={this.onWindowResize}
+                    mousedown={this.hide}
+                    touchstart={this.hide}
+                />
+                <div className="tooltip">
+                    {children}
+                    <svg className="tail" height="5" width="10">
+                        <polygon points="0,0 10,0 5,5"/>
+                    </svg>
+                </div>
+            </div>
+        );
+    }
+
+    public componentDidMount() {
+        this.setTarget();
+        this.bindEvents();
+        this.setStyles();
+    }
+    public componentWillUnmount() {
+        window.clearTimeout(this.timeout!);
+        this.unbindEvents();
+        this.onWindowResize.clear();
+    }
+    public componentWillReceiveProps(props: TooltipProps) {
+        this.unbindEvents();
+        this.setTarget();
+        this.bindEvents();
+        this.setStyles();
+    }
```

Comment:
are you sure we can't save here something and we have to do all calculations on any prop change?

Reply:
we need to recalculate this when user change `id` or `position`. I will add this condition